quinn/src/lib.rs

```diff
@@ -330,50 +316,63 @@ impl EndpointInner {
         ipv6: bool,
     ) -> Self {
+        let (sender, events) = mpsc::unbounded();
         Self {
             log,
             socket,
             inner,
             incoming,
             ipv6,
-            outgoing: None,
-            pending: FnvHashMap::default(),
-            timers: FuturesUnordered::new(),
+            outgoing: VecDeque::new(),
             buffered_incoming: VecDeque::new(),
             driver: None,
+            connecting: FnvHashMap::default(),
+            connections: FnvHashMap::default(),
+            sender,
+            events,
         }
     }
 
-    /// Wake up a blocked `Driver` task to process I/O
-    fn notify(&self) {
-        if let Some(x) = self.driver.as_ref() {
-            x.notify();
+    fn drive_connecting(&mut self) {
+        for new in self.connecting.values_mut() {
+            new.poll().unwrap();
         }
     }
 
-    fn drive_recv(
-        &mut self,
-        now: Instant,
-        wrapped: Rc<RefCell<EndpointInner>>,
-    ) -> Result<bool, io::Error> {
+    fn drive_recv(&mut self, now: Instant) -> Result<bool, io::Error> {
         let mut buf = [0; 64 * 1024];
         let mut recvd = 0;
         loop {
             match self.socket.poll_recv(&mut buf) {
                 Ok(Async::Ready((n, addr, ecn))) => {
-                    if let Some(ch) = self.inner.handle(now, addr, ecn, (&buf[0..n]).into()) {
-                        self.pending.insert(ch, Pending::new(None));
-                        match self.incoming.poll_ready() {
-                            Ok(Async::Ready(())) => {
-                                self.incoming
-                                    .start_send(NewConnection::new(wrapped.clone(), ch))
-                                    .unwrap();
-                                self.inner.accept();
+                    match self.inner.handle(now, addr, ecn, (&buf[0..n]).into()) {
+                        Some((handle, ConnectionEvent::NewConnection(conn))) => {
+                            let conn = self.create_connection(handle, conn, None);
+                            let conn = Rc::new(RefCell::new(conn));
+                            match self.incoming.poll_ready() {
+                                Ok(Async::Ready(())) => {
+                                    self.incoming
+                                        .start_send(NewConnection {
+                                            driver: ConnectionDriver(conn.clone()),
+                                            connection: Connection(conn.clone()),
+                                            incoming: IncomingStreams(conn),
+                                        })
+                                        .unwrap();
+                                    self.inner.accept();
+                                }
+                                _ => {
+                                    self.buffered_incoming.push_back((handle, Connection(conn)));
+                                }
```

Comment:
Pre-existing issue (i.e. not a blocker), but it would probably be a lot nicer to just unconditionally buffer everything here, and let `drive_incoming` worry about the rest. So long as people aren't setting ridiculously huge accept buffer sizes, anyway.

Reply:
Yes, that is nicer -- done.